internal/servers/controller/handlers/credentiallibraries/credentiallibrary_service.go

```diff
@@ -0,0 +1,602 @@
+func validateGetRequest(req *pbs.GetCredentialLibraryRequest) error {
+	return handlers.ValidateGetRequest(handlers.NoopValidatorFn, req, vault.CredentialLibraryPrefix)
+}
+
+func validateCreateRequest(req *pbs.CreateCredentialLibraryRequest) error {
+	return handlers.ValidateCreateRequest(req.GetItem(), func() map[string]string {
+		badFields := map[string]string{}
+		if !handlers.ValidId(handlers.Id(req.GetItem().GetCredentialStoreId()), vault.CredentialStorePrefix) {
+			badFields[globals.CredentialStoreIdField] = "This field must be a valid project scope id."
+		}
+		switch credential.SubtypeFromType(req.GetItem().GetType()) {
+		case credential.VaultSubtype:
+			attrs := &pb.VaultCredentialLibraryAttributes{}
+			if err := handlers.StructToProto(req.GetItem().GetAttributes(), attrs); err != nil {
+				badFields[globals.AttributesField] = "Attribute fields do not match the expected format."
+				break
+			}
+			if attrs.VaultPath == "" {
+				badFields[vaultPathField] = "This is a required field."
+			}
+			if m := attrs.GetHttpMethod(); m != nil && !strutil.StrListContains([]string{"GET", "POST"}, strings.ToUpper(m.GetValue())) {
+				badFields[httpMethodField] = "If set, value must be 'GET' or 'POST'."
+			}
+			if b := attrs.GetHttpRequestBody(); b != nil && strings.ToUpper(attrs.GetHttpMethod().GetValue()) != "POST" {
+				badFields[httpRequestBodyField] = fmt.Sprintf("Field can only be set if %q is set to the value 'POST'.", httpMethodField)
+			}
+		default:
+			badFields[globals.TypeField] = "This is a required field and must be a known credential library type."
+		}
+		return badFields
+	})
+}
+
+func validateUpdateRequest(req *pbs.UpdateCredentialLibraryRequest) error {
+	return handlers.ValidateUpdateRequest(req, req.GetItem(), func() map[string]string {
+		badFields := map[string]string{}
+		switch credential.SubtypeFromId(req.GetId()) {
+		case credential.VaultSubtype:
+			if req.GetItem().GetType() != "" && credential.SubtypeFromType(req.GetItem().GetType()) != credential.VaultSubtype {
+				badFields[globals.TypeField] = "Cannot modify resource type."
+			}
+			attrs := &pb.VaultCredentialLibraryAttributes{}
+			if err := handlers.StructToProto(req.GetItem().GetAttributes(), attrs); err != nil {
+				badFields[globals.AttributesField] = "Attribute fields do not match the expected format."
+				break
+			}
+			if handlers.MaskContains(req.GetUpdateMask().GetPaths(), vaultPathField) && attrs.VaultPath == "" {
+				badFields[vaultPathField] = "This is a required field and cannot be set to empty."
+			}
+			if m := attrs.GetHttpMethod(); handlers.MaskContains(req.GetUpdateMask().GetPaths(), httpMethodField) && m != nil && !strutil.StrListContains([]string{"GET", "POST"}, strings.ToUpper(m.GetValue())) {
+				badFields[httpMethodField] = "If set, value must be 'GET' or 'POST'."
+			}
+			if b := attrs.GetHttpRequestBody(); b != nil && strings.ToUpper(attrs.GetHttpMethod().GetValue()) == "GET" {
```

Comment:
There are a couple of options here...

As-is, if you are taking in a raw request body, you should validate that it can be parsed successfully as JSON. This may not be possible anymore at some point if we allow templating, which...we probably will...but even in that case likely it'd be specific values that are templated and they could just be properly JSON-escaped.

The other option here is to take in a map[string]string, then do that formatting into JSON ourselves at request time. This has the advantage that it's more UX-friendly -- you're defining the parameters of the request body without worrying about the JSON formatting and escaping. So it's easier on the CLI and in the UI. Generally speaking Vault's API is designed to be able to convert from string to necessary other types depending on context (e.g. if an endpoint needs an int for a param, it will do that translation for you) but there are a few places here and there where a raw value might be expected/needed, in which case you'd need a raw request body parameter anyways.

Of course, a third option is do the nicer UX thing first, then add a raw request body parameter later _if needed_ and make those mutually exclusive.

Probably worth roping in opinions from FE and PM.